docs/MockFunctionAPI.md

```diff
@@ -594,6 +608,39 @@ test('returns correct data', () => {
 
 Types of classes, functions or objects can be passed as type argument to `jest.Mocked<Source>`. If you prefer to constrain the input type, use: `jest.MockedClass<Source>`, `jest.MockedFunction<Source>` or `jest.MockedObject<Source>`.
 
+### `jest.Replaced<Source>`
+
+The `jest.Replaced<Source>` utility type returns the `Source` type wrapped with type definitions of Jest [replaced property](#replaced-properties).
+
+```ts title="src/utils.ts"
+export function isLocalhost(): boolean {
+  return process.env['HOSTNAME'] === 'localhost';
+}
+```
+
+```ts
+import {afterEach, expect, it, jest} from '@jest/globals';
+import {isLocalhost} from '../utils';
+
+let replacedEnv: jest.Replaced<typeof process.env> | undefined = undefined;
+
+afterEach(() => {
+  replacedEnv?.restore();
+});
+
+it('isLocalhost should detect localhost environment', () => {
+  replacedEnv = jest.replaceProperty(process, 'env', {HOSTNAME: 'localhost'});
```

Comment:
question: Why do we replace all of `process.env` instead of replacing only the value we're interested in i.e. why not
```suggestion
  replacedEnv = jest.replaceProperty(process.env, 'HOSTNAME', 'localhost');
```
 instead?

Reply:
yeah, that seems better, although then `replacedEnv` seems wrong?

Reply:
This opens a design question. One cannot right now replace property that does not exist on an object. Assuming 'HOSTNAME' might or might not be in the env... What should we do in this case?

The Sinon.js implementation forbids creation of new properties. It must exist on an object (or in the prototype chain). I applied same restriction here.

What do you think, @SimenB , @mrazauskas and @eps1lon ? Should we allow adding undefined properties to the object (which is actually quite simple to do), or should be more restrictive (and possibly open this behavior in future). I do not have strong thoughts here...

Reply:
> One cannot right now replace property that does not exist on an object. Assuming 'HOSTNAME' might or might not be in the env... What should we do in this case?

Hm good point. I would expect that we can mock non-existing properties. But can see how `replaceProperty` might be misleading. But again, from my intuition, we can mock non-existing properties.

Reply:
we could have a separate options bag (`{mustExist: true}` by default or something) that allows customization? But I think by default we should _not_ allow it - if nothing else to catch typos (TS might help of course).

Reply:
Yeah that works for our use case. And the `process.env` use case probably also works with a `process.env.HOST_NAME = undefined` during setup since I haven't seen any use case where you differentiate between unknown property and `undefined` in `process.env`

Reply:
> But again, from my intuition, we should be able to mock non-existing properties.

@eps1lon: I am a bit afraid of typos or invalid usages. People might be mocking stuff:
* that is not used due to typo,
* that exists now, but after upgrade of a library disappears (e.g. mocking private handle of some database client instance),
* due to error - misinterpreting API documentation and trying to mock property for different class.
 
And if one thinks about it... How common/rare is the use case when we need to mock something that does not exist? I can think of a case when people might want to mock either a new key of "dictionary" or some new array element. In that case, they should mock the parent anyway I guess... In same manner like the process.env. E.g.:

```js
// Instead of
jest.replaceProperty(httpAgent.defaultHeaders, 'X-API-Key', mockKey);

// Use
jest.replaceProperty(httpAgent, 'defaultHeaders', {...httpAgent.defaultHeaders, 'X-API-Key': mockKey});
```

* ❓ If we keep current behavior, should I mention in documentation how to work with undefined propeties - e.g. in process.env? This might be quite common case. 

---

> ...separate options bag (`{mustExist: true}` by default... by default we should not allow it - if nothing else to catch typos (TS might help of course).

@SimenB This makes sense. How should I proceed?

* ❓ Should I implement the optional ability to replace undeclared properties? 
* If yes, what is best name for the option inside property bag? (`mustExist: true` / `allowUndefined: false` / `allowNonexistent: false` / ...? I am fine with `mustExist`)

Reply:
Thanks, this makes sense. I will implement it as separate PR, as this is already growing quite a lot. :)